text/0692-deprecate-array-observers.md

```diff
@@ -0,0 +1,442 @@
+    }
+  },
+});
+```
+
+Ideally the dirty state would be converted into derived state that could read
+the array it was dependent upon, but if that's not an option or would require
+major refactors, it is also possible to override the mutator method of the array
+and trigger the change when it is called. In EmberArray's, the primary mutator
+method is the `replace()` method.
+
+```js
+const StatefulArray = ArrayProxy.extend(Copyable, {
+  content: computed(function() {
+    return A();
+  }),
+
+  // ...
+
+  replace() {
+    this._super(...arguments);
+
+    let record = get(this, 'owner');
+    let key = get(this, 'name');
+
+    // Any change to the size of the fragment array means a potential state change
+    if (get(this, 'hasDirtyAttributes')) {
+      fragmentDidDirty(record, key, this);
+    } else {
+      fragmentDidReset(record, key);
+    }
+  },
+});
+```
+
+Note that this method will work for arrays and array proxies that are mutated
+directly, but will not work for array proxies which wrap other arrays and watch
+changes on them. In those cases, the recommendation is to refactor such that:
+
+1. Changes are always intercepted by the proxy, and can call the code
+   synchronously when they occur.
+2. The change logic is added by intercepting changes on the original array, so
+   it will occur whenever it changes.
+3. The API that must be called synchronously is instead driven by derived state.
+   For instance, in the example above, the record's dirty state could be driven
+   by the various child fragments it contains, and updated whenever the user
+   accesses it, rather than by sending events such as `didDirty` and `didReset`.
+
+### Converting code that uses the `willChange` functionality
+
+In general, it is no longer possible to react to an array change before it
+occurs except by overriding the mutation methods on the array itself. You can do
+this by replacing them and calling your logic _before_ calling `super`.
+
+```js
+const ArrayWithWillChange = EmberObject.extend(MutableArray, {
+  replace() {
+    // Your logic here
+
+    this._super(...arguments);
+  },
+});
+```
+
+In cases where this is not possible, you can instead convert to derived state,
+and cache the previous value of the array to compare it the next time the state
+is accessed.
+
+## How We Teach This
+
+Array observers and their usage in general is not part of the default Ember
+learning path, so the main guides should not need to change. Eventually, we will
+likely want to add some guides for developing custom iterables that work in a
+lazy way, but this is a separate change that should not be tied to this specific
+deprecation.
+
+For the deprecation guide, see the transition path above.
+
+## Drawbacks
+
+- While array observers are not used commonly in public code, they are used by
+  several commonly used addons, such as `ember-collection`,
+  `ember-data-model-fragments`, and `ember-light-table`. Deprecating them will
```

Comment:
Do we want to mention ember-data here?  Seems like this might be an intermediate refactor for hasMany objects.

https://github.com/emberjs/data/blob/623145d0009d683a50468ba9961431ae76ae6df6/packages/model/addon/-private/system/many-array.js#L164

Reply:
Calling `arrayContentWillChange` and `arrayContentDidChange` is how an EmberArray _triggers_ array observers. Internally, Ember Data does not rely on array observers at all, but it must call these methods in case _others_ are using them.

This is a good point though, in that it's a similar situation to the one we're facing with implicit injections, Ember Data cannot remove these method calls until there are no more users, which means they cannot immediately comply with a deprecation. So, I think we should make it a phased deprecation, where `arrayContentWillChange` and `arrayContentDidChange` are fully deprecated in the next major version after array observers themselves are deprecated and removed.